include/deal.II/base/patterns_tools.h

```diff
@@ -0,0 +1,682 @@
+{
+  namespace internal
+  {
+    /**
+     * Store information about the rank types of the given class.
+     *
+     * A class has Rank equal to the number of different separators
+     * that are required to uniquely identify its element(s) in a string.
+     *
+     * This class is used to detect wether the class T is compatible
+     * with a Patterns::List pattern or with a Patterns::Map pattern.
+     *
+     * Objects like Point<dim> or std::complex<double> are vector-likes, and
+     * have vector_rank 1. Elementary types, like `int`, `unsigned int`, `double`, etc.
+     * have vector_rank 0. `std::vector`, `std::list` and in general containers have rank
+     * equal to 1 + vector_rank of the contained type. Similarly for map types.
+     *
+     * A class with list_rank::value = 0 is either elementary or a
+     * map. A class with map_rank::value = 0 is either a List compatible
+     * class, or an elementary type.
+     *
+     * Elementary types are not compatible with Patterns::List, but
+     * non elementary types, like Point<dim>(), or std::complex<double>, are
+     * compatible with the List type. Adding more compatible types is a matter
+     * of adding a specialization of this struct for the given type.
+     *
+     * @author Luca Heltai, 2017
+     */
+    template<class T, class Enable=void>
+    struct RankInfo
+    {
+      static constexpr int list_rank = 0;
+      static constexpr int map_rank = 0;
+    };
+  }
+
+  // Arithmetic types
+  template<class T>
+  struct Convert<T, typename std::enable_if<std::is_arithmetic<T>::value>::type>
+  {
+
+    static std::unique_ptr<PatternBase> to_pattern()
+    {
+      if (std::is_integral<T>::value)
+        return std_cxx14::make_unique<Integer>(std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+      else if (std::is_floating_point<T>::value)
+        return std_cxx14::make_unique<Double>(-std::numeric_limits<T>::max(), std::numeric_limits<T>::max());
+      else
+        {
+          AssertThrow(false, ExcNotImplemented());
+          return std::unique_ptr<PatternBase>();
+        }
+    }
+
+    static std::string to_string(const T &value, const std::unique_ptr<PatternBase> &p = Convert<T>::to_pattern())
+    {
+      std::string str;
+      if (std::is_same<T, unsigned char>())
+        str = std::to_string((int)value);
+      else
+        str = std::to_string(value);
+      AssertThrow(p->match(str), ExcNoMatch(str, *p));
+      return str;
+    }
+
+    static T to_value(const std::string &s,
+                      const std::unique_ptr<PatternBase> &p = Convert<T>::to_pattern())
+    {
+      AssertThrow(p->match(s), ExcNoMatch(s, *p));
+      std::istringstream is(s);
+      T value;
+      if (std::is_same<T,unsigned char>::value)
+        {
+          int i;
+          is >> i;
+          value=i;
+        }
+      else
+        is >> value;
+      AssertThrow(!is.fail(), ExcMessage("Failed to convert from \"" + s + "\" to the type \""
```

Comment:
Good point. Can you add a comment to this effect, referencing the case I mentioned? Since it wasn't clear to me, it may not be clear to the next reader either.